csrc/attention/attention_kernels.cu

```diff
@@ -261,6 +263,14 @@ __global__ void single_query_cached_kv_attention_kernel(
       if (row_idx < HEAD_SIZE) {
         const int offset = row_idx * BLOCK_SIZE + physical_block_offset;
         V_vec v_vec = *reinterpret_cast<const V_vec*>(v_ptr + offset);
+        // NOTE(woosuk): When v_vec contains the tokens that are out of the context,
+        // we should explicitly zero out the values since they may contain NaNs.
+        // See https://github.com/vllm-project/vllm/issues/641#issuecomment-1682544472
+        scalar_t* v_vec_ptr = reinterpret_cast<scalar_t*>(&v_vec);
+#pragma unroll
+        for (int j = 0; j <= V_VEC_SIZE; j++) {
+          v_vec_ptr[j] = token_idx + j < context_len ? v_vec_ptr[j] : zero_value;
+        }
```

Comment:
Good idea. I benchmarked the kernel.

* Before this PR:
```
torch.float16: 8 seqs, 30 context len, 8.6 us
torch.float16: 8 seqs, 498 context len, 52.5 us
torch.float16: 8 seqs, 2033 context len, 179.2 us
torch.float16: 8 seqs, 8168 context len, 678.8 us
torch.float16: 32 seqs, 30 context len, 19.6 us
torch.float16: 32 seqs, 498 context len, 180.2 us
torch.float16: 32 seqs, 2033 context len, 663.7 us
torch.float16: 32 seqs, 8168 context len, 2641.2 us
torch.float16: 128 seqs, 30 context len, 67.9 us
torch.float16: 128 seqs, 498 context len, 686.7 us
torch.float16: 128 seqs, 2033 context len, 2682.7 us
torch.float16: 128 seqs, 8168 context len, 10954.4 us
```

* Before adding `if(block_idx == num_blocks -1):
```
torch.float16: 8 seqs, 30 context len, 9.1 us
torch.float16: 8 seqs, 498 context len, 53.0 us
torch.float16: 8 seqs, 2033 context len, 178.5 us
torch.float16: 8 seqs, 8168 context len, 683.0 us
torch.float16: 32 seqs, 30 context len, 20.6 us
torch.float16: 32 seqs, 498 context len, 183.4 us
torch.float16: 32 seqs, 2033 context len, 668.3 us
torch.float16: 32 seqs, 8168 context len, 2674.8 us
torch.float16: 128 seqs, 30 context len, 70.9 us
torch.float16: 128 seqs, 498 context len, 698.1 us
torch.float16: 128 seqs, 2033 context len, 2706.8 us
torch.float16: 128 seqs, 8168 context len, 11026.7 us
```

* After adding `if(block_idx == num_blocks -1):
```
torch.float16: 8 seqs, 30 context len, 9.1 us
torch.float16: 8 seqs, 498 context len, 52.7 us
torch.float16: 8 seqs, 2033 context len, 181.4 us
torch.float16: 8 seqs, 8168 context len, 685.7 us
torch.float16: 32 seqs, 30 context len, 20.1 us
torch.float16: 32 seqs, 498 context len, 180.9 us
torch.float16: 32 seqs, 2033 context len, 664.7 us
torch.float16: 32 seqs, 8168 context len, 2661.6 us
torch.float16: 128 seqs, 30 context len, 73.2 us
torch.float16: 128 seqs, 498 context len, 694.1 us
torch.float16: 128 seqs, 2033 context len, 2690.4 us
torch.float16: 128 seqs, 8168 context len, 10963.3 us
```

Adding the if statement slightly improves the performance when the number of blocks is large. In general, I believe the overhead (which is < 5%) is now acceptable.

Reply:
Sounds good! Let's just add this if statement since it looks like a pure performance optimization?

Reply:
I feel bad for adding a condition in the inner loop.
Have you compared it with the approach of simply set the page to zeros when releasing a page?

Reply:
As I mentioned here. Padding seems more reasonable https://github.com/vllm-project/vllm/issues/641#issuecomment-1682619534

Reply:
Hi @Atry, thanks for the suggestion. I considered your solution that does not modify the krenel. While it is simple and never affects the performance, I felt handling NANs in the kernel would be a better design (as long as the impact on the performance is acceptable). The reason is, changing the kernel can 1) benefit all the kernel users besides vLLM (e.g., TGI), and 2) guarantee that previous requests cannot affect the future requests (e.g., when previous requests might somehow store NaNs in the KV cache).

Reply:
I suggest to fill the page with finite values when "releasing" a page. Because the CUDA operator is non-blocking, it would not affect the latency.

Another idea is to change the shape of v_cache like k_cache's, so that we can use the same approach to filter out zero values like k_cache. In fact, because we can skip some loops, I expect it could have a chance improve the performance if we change the shape of v_cache.